### PR TITLE
Fix ModelCheckpoint misformats filename

### LINKED ISSUE #16290: ModelCheckpoint misformats filename if keys have shared substrings
### Bug description

Using a filename such as this causes `ModelCheckpoint._format_checkpoint_name()` to replace shared substrings with `0[shared_substr]` in all keys with the substring.
```
    filename = "{epoch:04d}_{step:010d}_{epoch_test:.3g}"
```

In this case, the filename gets mangled to the following before metric values are inserted:
```
{0[epoch]:04d}_{0[step]:010d}_{0[epoch]_test:.3g}
```

The bug happens with and without `auto_insert_metric_name`.

### How to reproduce the bug

_No response_

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 1.10):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

### More info

I have a pretty simple patch. Happy to open a PR if this makes sense:

```python
    @classmethod
    def _format_checkpoint_name(
        cls,
        filename: Optional[str],
        metrics: Dict[str, Tensor],
        prefix: str = "",
        auto_insert_metric_name: bool = True,
    ) -> str:
        if not filename:
            # filename is not set, use default name
            filename = "{epoch}" + cls.CHECKPOINT_JOIN_CHAR + "{step}"

        # check and parse user passed keys in the string
        groups = re.findall(r"{([^\:}]*)(:?[^}]*)}", filename)
        if len(groups) >= 0:
            for name, fmt in groups:
                if auto_insert_metric_name:
                    # {name:fmt} to name={name:fmt}
                    metric_prefix = name + "="
                else:
                    metric_prefix = ""

                # 0[name] <- support for dots: https://stackoverflow.com/a/7934969
                old_substr = "{" + name + fmt + "}"
                new_substr = metric_prefix + "{0[" + name + "]" + fmt + "}"
                filename = filename.replace(old_substr, new_substr)

                if name not in metrics:
                    metrics[name] = torch.tensor(0)
            filename = filename.format(metrics)

        if prefix:
            filename = cls.CHECKPOINT_JOIN_CHAR.join([prefix, filename])

        return filename

```

cc @carmocca @awaelchli

### PR DESCRIPTION
**Pull request by bkiat1123**
## What does this PR do?

As reported in [this issue](https://github.com/Lightning-AI/lightning/issues/16290), filename format failed when one metric key is substring of another. 

```python
filename = "{epoch:03d}-{epoch_test:03d}"
groups = re.findall(r"(\{.*?)[:\}]", filename)
for group in groups:
    filename = filename.replace(group, ...)
````

In the first iteration of the for loop, both `epoch` and `epoch_test` will be formatted, which cause the error. Solve by reverse sorting the keys by length.

Fixes #16290 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Review by Borda — APPROVED** [review]

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -514,19 +514,23 @@ def _format_checkpoint_name(
...
         groups = re.findall(r"(\{.*?)[:\}]", filename)
-        if len(groups) >= 0:
-            for group in groups:
-                name = group[1:]
 
-                if auto_insert_metric_name:
-                    filename = filename.replace(group, name + "={" + name)
+        # sort keys from longest to shortest to avoid replacing substring
+        # eg: if keys are "epoch" and "epoch_test", the latter must be replaced first
+        groups = sorted(groups, key=lambda x: len(x), reverse=True)
```
Comment (awaelchli) [inline src/lightning/pytorch/callbacks/model_checkpoint.py:520]: Thanks for this comment. Now I finally get what root problem with the reported bug is 😅

File: src/lightning/pytorch/callbacks/model_checkpoint.py [hunk 2]
```diff
@@ -514,19 +514,23 @@ def _format_checkpoint_name(
 
         # check and parse user passed keys in the string
         groups = re.findall(r"(\{.*?)[:\}]", filename)
-        if len(groups) >= 0:
```
Comment (awaelchli) [inline src/lightning/pytorch/callbacks/model_checkpoint.py:517]: isn't this needed anymore and why?

**Review by awaelchli — APPROVED** [review]

File: src/lightning/pytorch/callbacks/model_checkpoint.py (hunk 2 above)
Comment (bkiat1123) [reply src/lightning/pytorch/callbacks/model_checkpoint.py:517]: Since groups is always a list, `if len(groups) >= 0` is equivalent to `if True`.

I have thought about changing to `if len(groups) > 0` instead. However, both the for loop and `str.format` works well when group is 0. If there is no curly bracket in the filename, the `format` method has no effect.

```python
filename = "best_model"  # groups = [] in this case
assert filename.format({})  == "best_model"
assert filename.format({"a": 1}) == "best_model"
```
